VectorSeries

VisualMelon opened the pull request:
Working on #107, putting this here so that anyone who wants to explain all the things wrong with it can do so, because I've not touched it recently and am making no progress, and would appreciate any feedback. There are also a couple of things in here I find myself wanting to point at on occasion.

### Checklist

- [x] I have included examples or tests
- [ ] I have updated the change log
- [x] I am listed in the CONTRIBUTORS file
- [ ] I have cleaned up the commit history (use rebase and squash)

### Changes proposed in this pull request:

- Add a `VectorSeries`, which plots a series of `VectorItems` as arrows on an X/Y axis pair
   - `VectorItems` have an `Origin`, `Direction`, and `Value` (i.e. it's a vector with some arbitrary value positioned at a particular point)
   - The value is mapped to a color on a `ColorAxis`
- Adds a `DataVector` struct
  - I would like to go through the existing code base and replace usage of `DataPoint` as a vector with `DataVector`, though I'm not sure there are any instances (I need to check)

I'm not entirely happy with rendering, but that might be because I'm never entirely happy.

It doesn't make plotting a vector field especially easy, because it just plots arrows. We could either provider helpers to map a vector field (i.e. 2D array of vectors) to a list of `VectorItems`, or we could make `VectorItem` optionally read from a multidimensional array, or we could provide a separate `VectorFieldSeries` to do that.

I opted for X/Y vector rather than Angle/Mag because the former would be confusing wrt. the plot aspect ratio: if a user wants angle/mag, then they should be fixing the aspect ratio. (Note: that isn't easy...)

Have I totally missed the point?

@oxyplot/admins

VisualMelon committed:
Vector stuff

VisualMelon committed:
Examples

VisualMelon committed:
Can't remember

VisualMelon committed:
Fix everything

Jonarw commented:
I'm not sure I can completely get behind the idea of defining the vectors using X/Y coordinates rather than Magnitude/Angle. This might have to do with the only kind of vector fields I ever really was in contact with - which is optical representations of magnetic/electric fields, something like [here](https://pages.vassar.edu/magnes/advanced-electromagnetism-phys-341/rachel-wagner-kaiser/).

Here it doesn't make sense to specify the vector in terms of X/Y coordinates, because it represents a completely different thing: while the X/Y coordinates represent a point in 2-dimensional space, the vector length and direction represents the field strength and direction. Consequently it also doesn't make sense to change the direction or length of the vector when the X or Y axes are scaled. Also in this context it makes no sense for the vector to become curved on logarithmic axes.

Of course there might be other applications I am unaware of where the vector is actually directly related to the X and Y coordinates - did you have such an application in mind?

VisualMelon commented:
@Jonarw thanks for taking a look. I don't have any particular application in mind (I did 3  or 4 years ago, but I can't remember what; I threw something together at the time and moved on).

I completely understand your point about the magnitudes being meaningful. The vector field with angle/mag only makes sense on a set of well aligned and scaled axes, in which case the representation doesn't matter: a vector is just a vector, and the user could use a `FromAngleAndMagnitude` method to create them in if suits their use case and enforce a pair of well aligned and scaled axes.
 - With an angle/mag representation, the magnitude is always meaningful, but the angle will be misleading on a pair of mis-aligned/scaled axes.
 - With a dx/dy representation, the direction is always meaningful, but the magnitude will be misleading on a pair of mis-scaled axis

On the other hand, a plot of some arbitrary differential function doesn't necessarily warrant an angle/magnitude interpretation at all. If interpretability is a concern then there should be a constraint on the axes to enforce interpretability, regardless of the representation.

Regarding X/Y being coordinates, this is why I introduce the `DataVector` type, which is explicitly _not_ a point, but rather a vector. The same has existed for ages with `ScreenPoint`/`ScreenVector`.

Regarding curves on log/other axis, I mostly implemented this for fun and to explore alternative to the current algorithm in `LineAnnotation`. It doesn't make sense with an angle/mag interpretation, but I don't know that an angle/mag interpretation could ever make sense on a pair of non-linear 2D axes.

To sum up, I think that either representation is problematic on mis-aligned/scaled axes, which leads to the conclusion that we should have a more powerful constraints mechanism to enforce consistency between axes. It's very hard to enforce constraints at the moment beyond the `Cartesian`: we should this as an extensibility point. In my opinion, x/y is the more natural representation of a vector, but I'm willing to be convinced otherwise.

Could it make sense to have 2 series?
 - one with a data-space vector representation (would it be better to just have a more general purpose stream plot?)
 - one with a angle/mag vector representation

Jonarw commented:
Hmm yes I see your point. I'll have to think about it some more.
In the meantime a (possibly stupid) idea: How about the the vector was specified by magnitude and angle, where the magnitude would be specified in screen space (and therefore unaffected by axes) and the angle in data space, so it would still be meaningful when the axes not cartesian. This would in theory solve the problem, but at the moment I am not sure if it wouldn't be too confusing or if makes any sense at all.

VisualMelon commented:
@Jonarw I might change it so that it is angle/mag, but the mag is data/screen space configurable to see how that feels. I really must take another look at matlab/matplotlib to see what people expect as well.

cliff-lane commented:
Has there been any more thought put into this project?   I am in need of a plotting library that includes Contour Plots, Histograms, and Vector Fields.  It appears that OxyPlot has everything I need with the exception of the finalized Vector Fields.  The data I need to plot is basically misalignment data where I have a (x,y) of where the point should be and a (x,y) of how far off the point is from where it should be.

VisualMelon commented:
Sorry, no: not worked on this at all.

For your use-case, sounds like it would make sense to have a dx/dy vector description, which is what the draft currently implements. You should be able to mostly C&P the implementation if you want to mess around with it: there's a lot of junk in the PR that shouldn't be there and can be ignored.

VisualMelon commented:
Sorry, no: not worked on this at all.

For your use-case, sounds like it would make sense to have a dx/dy vector description, which is what the draft currently implements. You should be able to mostly C&P the implementation if you want to mess around with it: there's a lot of junk in the PR that shouldn't be there and can be ignored.

tylerdoboy commented:
Will this become available before next year? I have been implementing my own vector field but it would be nice to have this series available.

VisualMelon commented:
@tylerdoboy seems there is a fair bit of interest in a vector series, so that'll push it higher in my list of things TODO if I have time.

Please do share any details you can about your use-case: would the design here work for you? Are there any features you need that are missing? etc. this will help with the design and let us know when we have something worth including in the library.

VisualMelon committed:
Fix DefaultTrackerFormatString

VisualMelon committed:
Tidy

VisualMelon committed:
Add slightly misleading MinimumSegmentLength

VisualMelon committed:
Adjustable vector positioning

VisualMelon commented:
Rebased, tidied a bit, and added the ability to position vector arrows/text across along their length to support a vector-field scenario.

If anyone has any fun examples they'd like to see implemented to ensure they have good support, let me know.

VisualMelon committed:
VectorSeries.ArrowHeadPosition

VisualMelon commented:
Do we need a nice means for plotting a 2D vector field? (current API demands you provide it with per-vector `VectorItem`s.

This could be a helper that generates a list of points from the 2D array, or it could be an alternative to `Items` in the series itself, or it could be implemented as a special ItemSource.

VisualMelon commented:
Definitely need to do some tidying here, and add a more sensible example, but reasonably happy with the API as is. Need to review the data-bounds logic as well.

VisualMelon committed:
Tidy examples

VisualMelon commented:
Current data-range is defined by the 'Origin' of each `VectorItem`; this means that the arrows may be clipped. We could change this so that it's the min/max point where a rendered arrow starts or stops, but I'm not completely sold on that idea.

Jonarw commented on Source/OxyPlot/Foundation/DataPoint.cs:
```diff
@@ -93,6 +93,98 @@ public bool Equals(DataPoint other)
+            intersection = line1Start + line1Dir * lambda1;
+            return lambda0 >= 0 && lambda0 <= 1 && lambda1 >= 0 && lambda1 <= 1;
+        }
+
```
I'm not sure if these function really belong in the `DataPoint` class - maybe they would be better suited in a separate static helper class? I am not sure if we already have one where this could fit, it not maybe we could add one?

Jonarw commented on Source/OxyPlot/Foundation/DataVector.cs:
```diff
@@ -0,0 +1,199 @@
+        /// </summary>
+        [SuppressMessage("StyleCop.CSharp.NamingRules", "SA1307:AccessibleFieldsMustBeginWithUpperCaseLetter", Justification = "Reviewed. Suppression is OK here.")]
+        // ReSharper disable once InconsistentNaming
+        internal double y;
```
Do we need these to be accessible internally? Couldn't they just be private?

VisualMelon replied:
Will leave as is for now; another thing we could change throughout the code-base. Certainly no point in having them if we want the structs to be immutable.

Jonarw commented on Source/OxyPlot/Foundation/DataVector.cs:
```diff
@@ -0,0 +1,199 @@
+            return new DataVector(-v.x, -v.y);
+        }
+
+        /// <summary>
+        /// Normalizes this vector inplace.
+        /// </summary>
+        /// <remarks>
+        /// If the vector has zero length, then it is left unchanged.
+        /// </remarks>
+        public void Normalize()
+        {
+            double l = Math.Sqrt((this.x * this.x) + (this.y * this.y));
+            if (l > 0)
+            {
+                this.x /= l;
+                this.y /= l;
+            }
+        }
```
I always have a bad feeling about mutable `struct`s. I think I would like it more if this returned a new struct.
Also this should use the `Length` property I reckon.

Jonarw commented on Source/OxyPlot/Foundation/DataVector.cs:
```diff
@@ -0,0 +1,199 @@
+#pragma warning disable 1718
+            // ReSharper disable EqualExpressionComparison
+            // ReSharper disable CompareOfFloatsByEqualityOperator
+            return this.x == this.x && this.y == this.y;
```
I am not quite sure if that comparison is actually faster than `double.NaN` (anymore?), however I am fairly sure that it wouldn't really matter in terms of magnitude even if there was a difference... so I think I would prefer `double.NaN` here to avoid all the surrounding clutter.

VisualMelon replied:
I completely agree: this is a hideous pattern and I can't believe it will be faster these days than `double.IsNaN`

Jonarw commented on Source/OxyPlot/Rendering/ScreenPoint.cs:
```diff
@@ -117,6 +117,63 @@ public static bool IsUndefined(ScreenPoint point)
+        /// <returns><c>true</c> if <paramref name="f"/> is non-NaN and non-Infinite, otherwise <c>false</c>.</returns>
+        private static bool IsFinite(double f)
+        {
+            return !(double.IsNaN(f) || double.IsInfinity(f));
```
Does this do something different than `double.IsFinite`?

VisualMelon replied:
I don't think so.

Jonarw commented on Source/OxyPlot/Rendering/ScreenPoint.cs:
```diff
@@ -117,6 +117,63 @@ public static bool IsUndefined(ScreenPoint point)
+        /// <param name="ray1direction">The direction of the second ray.</param>
+        /// <param name="lambda0">The distance along the first ray in multiples of its direction.</param>
+        /// <param name="lambda1">The distance along the second ray in multiples of its direction.</param>
+        /// <returns><c>True</c> if the rays intersect, otherwise <c>false</c></returns>
+        public static bool IntersectRayWithRay(ScreenPoint ray0start, ScreenVector ray0direction, ScreenPoint ray1start, ScreenVector ray1direction, out double lambda0, out double lambda1)
+        {
+            var numerator0 = ray1direction.X * (ray0start.Y - ray1start.Y) - ray1direction.Y * (ray0start.X - ray1start.X);
+            var numerator1 = ray0direction.X * (ray1start.Y - ray0start.Y) - ray0direction.Y * (ray1start.X - ray0start.X);
+            var denomonator = ray1direction.X * ray0direction.Y - ray0direction.X * ray1direction.Y;
+
+            lambda0 = numerator0 / denomonator;
+            lambda1 = numerator1 / denomonator;
+
+            return IsFinite(lambda0) && IsFinite(lambda1);
+        }
+
+        /// <summary>
+        /// Intersects two lines in screen space.
+        /// </summary>
+        /// <param name="line0Start">The start point of the first line.</param>
+        /// <param name="line0End">The end point of the first line.</param>
+        /// <param name="line1Start">The start point of the second line.</param>
+        /// <param name="line1End">The end point of the second line.</param>
+        /// <param name="intersection">The point where the two lines intersect.</param>
+        /// <returns><c>True</c> if the lines intersect, otherwise false.</returns>
+        public static bool IntersectLineWithLine(ScreenPoint line0Start, ScreenPoint line0End, ScreenPoint line1Start, ScreenPoint line1End, out ScreenPoint intersection)
+        {
+            var line0Dir = line0End - line0Start;
+            var line1Dir = line1End - line1Start;
+            bool notParallel = IntersectRayWithRay(line0Start, line0Dir, line0Start, line0Dir, out double lambda0, out double lambda1);
+
+            if (!notParallel)
+            {
+                intersection = default;
+                return false;
+            }
+
+            intersection = line1Start + line1Dir * lambda1;
+            return lambda0 >= 0 && lambda0 <= 1 && lambda1 >= 0 && lambda1 <= 1;
+        }
```
I don't really like the fact that these are duplicated between `ScreenPoint` and `DataPoint`, but I suppose the same could also be said about other functionality that is common to `ScreenPoint` and `DataPoint`. Maybe we should consider adding something like an `IPoint` interface in the future and implement stuff like this as generic static functions. But I guess this might be out of scope of this PR, so I'm happy if it stays like this for now

VisualMelon replied:
These statics aren't actually being used; I'm not sure why I added them. I will remove them.

Jonarw commented on Source/OxyPlot/Rendering/ScreenVector.cs:
```diff
@@ -134,6 +134,9 @@ public double Y
+        /// <remarks>
+        /// If the vector has zero length, then it is left unchanged.
+        /// </remarks>
         public void Normalize()
```
I see that there is some precedent for the mutable `struct` stuff, and also for pretty much all the other stuff I didn't like in `DataVector`... And unfortunately, changing `ScreenVector.Normalize` to be non-mutable would be a breaking change... But maybe one we would be willing to take?

VisualMelon replied:
Yep, unfortunate history here. I think we should make `DataVector` immutable whether `ScreenVector` is or not (as you say, mutable struct semantics are ludicrously complicated), and probably mark this method as obsolete.

VisualMelon replied:
Removed the method in `DataVector` and made `x` and `y` readonly for the time being so there's no risk of not getting around to that once this goes in.

Jonarw commented on Source/OxyPlot/Series/VectorItem.cs:
```diff
@@ -0,0 +1,100 @@
+#pragma warning disable 1718
+            // ReSharper disable EqualExpressionComparison
+            // ReSharper disable CompareOfFloatsByEqualityOperator
+            return this.Origin.IsDefined() && this.Direction.IsDefined() && !double.IsNaN(this.Value);
```
I think the surrounding comments are not really applicable - Copy&Paste?

VisualMelon replied:
Fixed this in VectorItem and RectangleItem (another C&P job). Also changed the definition of `Equals` in both: this is a breaking change for `RectangleItem` but the old versions looks wrong to me.

Jonarw reviewed (commented):
Some smaller comments just from looking at the code changes, I will play around with the example later.

> Current data-range is defined by the 'Origin' of each VectorItem; this means that the arrows may be clipped. We could change this so that it's the min/max point where a rendered arrow starts or stops, but I'm not completely sold on that idea.

I guess arguments could be made for both, I am not quite sure yet on which side I am on this ;)

VisualMelon commented:
@Jonarw lots of comments here about `DataVector` that apply equally to `ScreenVector`: I'd suggest we handle those in a separate PR where we update both to be less hideous.

Jonarw commented:
> @Jonarw lots of comments here about `DataVector` that apply equally to `ScreenVector`: I'd suggest we handle those in a separate PR where we update both to be less hideous.

Yes, fine with me!

VisualMelon committed:
Address issues in VectorItem and equivalent concerns in RectangleItem

VisualMelon committed:
Make DataVector readonly

VisualMelon committed:
Fix DataVector to have internal fields (for consistency)

VisualMelon commented:
Regarding examples, they're pretty terrible at the moment, but I'm happy to implement any more exciting/useful ideas you might have.

VisualMelon committed:
Remove unused statics

VisualMelon committed:
Remove unused IsFinite static

VisualMelon committed:
Remove unused IsFinite static

VisualMelon commented:
I've marked the bits that are not vector-series specific as resolved for now to keep things clear for this PR; observations so far:
 - Would prefer everything is immutable
 - If everything is immutable, not point having `interval` fields for x/y
 - Normalise should use `Length` rather than computing it independently
 - Use `IsNaN` instead of `x == x` everywhere

Jonarw commented on Source/OxyPlot/Series/VectorSeries.cs:
```diff
@@ -0,0 +1,569 @@
+        /// Gets the list of Vectors.
+        /// </summary>
+        /// <value>A list of <see cref="VectorItem" />. This list is used if <see cref="ItemsSeries.ItemsSource" /> is not set.</value>
+        public List<VectorItem> Items { get; } = new List<VectorItem>();
```
I would prefer to expose an `IList<VectorItem>`. I know we expose `List`s elsewhere, so it would be somewhat inconsistent, but as far as I am concerned it is still preferable.

Jonarw commented on Source/OxyPlot/Series/VectorSeries.cs:
```diff
@@ -0,0 +1,569 @@
+        /// </summary>
+        /// <param name="rc">The rendering context.</param>
+        /// <param name="items">The Items to render.</param>
+        protected void RenderVectors(IRenderContext rc, ICollection<VectorItem> items)
```
`items` could be `IEnumerable<>`

Jonarw reviewed (approved):
I have two small, somwhat opinion-based, comments about the API, otherwise I think this is good to merge!

The Examples are sufficient to see that everything works as it should, so they fulfill their purpose I think.

VisualMelon committed:
Relax collection types

VisualMelon commented:
I agree with both of those, and relaxed the protected `ActualItems` also (should it be `IReadOnlyList`?): no reason we can't use more sensible types on new series, and if we later decide to change to the concrete type then it's not a code breaking change.

_**Edit**: can't be IReadOnlyList because IList doesn't implement it_

VisualMelon commented:
Forgot the changelog...

VisualMelon committed:
Update Changelog

VisualMelon commented:
Hopefully this is in a mergeable state, but I'd rather we didn't merge it just yet, as I'd like to give it another review myself in a couple of days when my head is clearer.

VisualMelon commented:
Rebased. I don't have any more changes I want to make now.

VisualMelon commented:
Rebased
